Parametrize `baseUrl` for GoogleAi Gemini models.

kpavlov opened the pull request:
## Issue
Currently it is not possible to integration-test Gemini models with mocks

## Change

Parametrize `baseUrl` for GoogleAi Gemini models.

- Refactor GoogleAiGemini models to support custom API `baseUrl`.
- Move `apiKey` to GeminiService, don't store it in model classes
- Add new error handling test for GoogleAiGeminiChatModel.
- Update `ai-mocks` dependency to version 0.4.3


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

kpavlov committed:
Parametrize `baseUrl` for GoogleAi Gemini models.

- Refactor GoogleAiGemini model to support customizable `baseUrl`.
- Move `apiKey` to GeminiService
- Add new error handling tests for GoogleAiGeminiChatModel.
- Update `ai-mocks` dependency to version 0.4.3 and

kpavlov committed:
Merge branch 'main' into gemini-base-url

glaforge reviewed (dismissed):
LGTM
Thanks for looking into this!

kpavlov committed:
Merge branch 'main' into gemini-base-url

glaforge committed:
Merge branch 'main' into gemini-base-url

dliubarskyi commented on langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java:
```diff
@@ -49,15 +49,15 @@ public GoogleAiEmbeddingModel(GoogleAiEmbeddingModelBuilder builder) {
     public GoogleAiEmbeddingModel(
             String modelName,
             String apiKey,
+            String baseUrl,
```
nit: That is a breaking change. Although it does not matter since this ctor was deprecated and will be removed soon. Just a heads up for the future

dliubarskyi commented on langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiEmbeddingModel.java:
```diff
@@ -165,15 +154,15 @@ public static class GoogleAiEmbeddingModelBuilder {
         private HttpClientBuilder httpClientBuilder;
         private String modelName;
         private String apiKey;
+        private String baseUrl = GEMINI_AI_ENDPOINT;
```
We usually assign default values in ctor, not in the builder. There are 2 benefits:
- all defaults in one place
- no way to accidentally override default value with null

kpavlov replied:
Thank you @dliubarskyi. Moved fallback to default to GeminiService

dliubarskyi commented on langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiChatModel.java:
```diff
@@ -200,6 +204,7 @@ public static class GoogleAiGeminiChatModelBuilder {
         private HttpClientBuilder httpClientBuilder;
         private ChatRequestParameters defaultRequestParameters;
         private String apiKey;
+        private String baseUrl = GEMINI_AI_ENDPOINT;
```
same here

dliubarskyi commented on langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiStreamingChatModel.java:
```diff
@@ -131,6 +133,7 @@ public static class GoogleAiGeminiStreamingChatModelBuilder {
         private HttpClientBuilder httpClientBuilder;
         private ChatRequestParameters defaultRequestParameters;
         private String apiKey;
+        private String baseUrl = GeminiService.GEMINI_AI_ENDPOINT;
```
and here

dliubarskyi commented on langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiTokenCountEstimator.java:
```diff
@@ -104,12 +103,13 @@ public static class Builder {
         private String modelName;
         private String apiKey;
+
+        private String baseUrl = GeminiService.GEMINI_AI_ENDPOINT;
```
and here

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiChatModel.java:
```diff
@@ -311,21 +311,12 @@ public VertexAiGeminiChatModel(
-        } else {
-            this.logResponses = false;
-        }
+        this.logRequests = Objects.requireNonNullElse(logRequests, false);
```
```suggestion
        this.logRequests = getOrDefault(logRequests, false);
```
To keep it consistent

dliubarskyi commented on langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiChatModel.java:
```diff
@@ -311,21 +311,12 @@ public VertexAiGeminiChatModel(
-            this.logResponses = false;
-        }
+        this.logRequests = Objects.requireNonNullElse(logRequests, false);
+        this.logResponses = Objects.requireNonNullElse(logResponses, false);
```
```suggestion
        this.logResponses = getOrDefault(logResponses, false);
```

dliubarskyi reviewed (commented):
@kpavlov thank you!

kpavlov committed:
Update langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiChatModel.java

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

kpavlov committed:
Update langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/gemini/VertexAiGeminiChatModel.java

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

kpavlov committed:
Merge branch 'main' into gemini-base-url

kpavlov committed:
Introduce `baseUrl` handling in GeminiService

- Don't assign default value of `baseUrl` in builders
- Check `baseUrl` for null and use default in GeminiService only.
- Move default timeouts to constants.
- Introduce Utils.firstNotNull(...)

kpavlov committed:
Merge branch 'main' into gemini-base-url

dliubarskyi reviewed (approved):
@kpavlov thanks a lot!
